### PR TITLE
feat: `pagerduty_service_integration_email` to trigger alerts on recieving email.

### PR DESCRIPTION
**Pull request by bhargavms**
Please confirm that you have done the following before requesting reviews:

- [x] I have confirmed that the PR type is appropriate for the change I am making according to the [Honest Pull Request and Commit Message Naming Conventions](https://www.notion.so/honestbank/Pull-Request-and-Commit-Message-Naming-Conventions-bd97f2cbb34c4c73b1ff3a3e384b850c).
- [x] I have typed an adequate description that explains **why** I am making this change.
- [x] I have installed and run standard pre-commit hooks that lints and validates my code.

### Description
This feature solves the requirement to automatically trigger alerts upon receiving specific emails to specific email addresses.

- - -
This change is [Reviewable](https://reviewable.io/reviews/honestbank/terraform-pagerduty-alerting/36)

### CODE REVIEW

**Comment by jai** [comment]
I think we may need a synchronous review

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -4,23 +4,32 @@ repos:
...
+        language: system
   - repo: https://github.com/gitguardian/ggshield
     rev: v1.14.4
     hooks:
       - id: ggshield
         language: python
         stages: [commit]
         args: [ 'secret', 'scan', 'pre-commit' ]
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.4.0 # Get the latest version from: https://github.com/pre-commit/pre-commit-hooks/releases
```
Comment (jai) [inline .pre-commit-config.yaml:32]: Is this our new standard?

File: examples/pagerduty-escalation-policy/main.tf [hunk 2]
```diff
@@ -1,38 +1,46 @@
+resource "random_id" "random_suffix" {
+  byte_length = 10
```
[the remark this answers is not in the document]
Comment (bhargavms) [reply examples/pagerduty-escalation-policy/main.tf:2]: I saw examples of byte length 10 being used in other projects that's why I went with the same, 4 should be good enough entropy I agree

File: README.md [hunk 3]
```diff
@@ -38,5 +38,34 @@ No resources.
 
 ## Outputs
 
+No outputs.
+<!-- END_TF_DOCS -->- BEGIN_TF_DOCS -->
```
[the remark this answers is not in the document]
Comment (bhargavms) [reply README.md:42]: make docs fails me :( will fix

File: .pre-commit-config.yaml (hunk 1 above)
Comment (bhargavms) [reply .pre-commit-config.yaml:32]: should be like this

https://github.com/honestbank/.github/blob/main/others/pre-commit-config-terraform.yaml

will update, but yes end of line and white space fixes has been moved to the  end

File: test/pagerduty_service_test.go [hunk 4]
```diff
@@ -45,7 +45,12 @@ func createPagerdutyService(t *testing.T, workingDir string, runID string) (stri
 }
 
 func destroyPagerdutyService(t *testing.T, workingDir string) {
-	terraform.Destroy(t, test_structure.LoadTerraformOptions(t, workingDir))
+	_, err := terraform.DestroyE(t, test_structure.LoadTerraformOptions(t, workingDir))
```
Comment (bhargavms) [inline test/pagerduty_service_test.go:48]: these tests feel like very easy to become flaky

**Review by MXfive — CHANGES_REQUESTED** [review]

**Review by MXfive — APPROVED** [review]

File: examples/pagerduty-business-service/main.tf [hunk 5]
```diff
@@ -1,15 +1,23 @@
...
+
+locals {
+  random_suffix = random_id.random_suffix.b64_url
+}
+
 module "mock_team" {
   source = "../../modules/pagerduty-team"
 
-  name        = "team - ${var.name}"
+  name        = "team - ${var.name} - ${local.random_suffix}"
```
Comment (jai) [inline examples/pagerduty-business-service/main.tf:12]: spacing in naming is inconsistent - for example in examples/honest-two-level-schedule/main.tf the name is:

```
  name        = "Example-${var.name}-${local.random_suffix}"
```

(no spacing around hyphens).

Not the end of the world, just a stylistic nit.

File: examples/pagerduty-service-integration-email/README.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+# PagerDuty Service Integrations - Email Example.
```
Comment (jai) [inline examples/pagerduty-service-integration-email/README.md:1]:
```suggestion
# PagerDuty Service Integrations - Email Example
```

File: examples/pagerduty-business-service/main.tf (hunk 5 above)
Comment (bhargavms) [reply examples/pagerduty-business-service/main.tf]: fixed in latest commit please re-review 🙏

File: examples/pagerduty-service/main.tf [hunk 7]
```diff
@@ -7,7 +11,7 @@ module "escalation_policy" {
 
 module "service" {
   source      = "../../modules/pagerduty-service"
-  name        = var.name
+  name        = "${var.name}-${random_id.random_suffix.b64_url}"
```
Comment (jai) [inline examples/pagerduty-service/main.tf:14]: Is this intentional? `b64_url`

File: examples/pagerduty-stakeholder/inputs.tf [hunk 8]
```diff
@@ -1,5 +1,5 @@
 variable "email_address" {
-  description = "The email adddress of the user."
+  description = "The email address of the user."
```
Comment (jai) [inline examples/pagerduty-stakeholder/inputs.tf:2]: nice catch :)

File: examples/pagerduty-business-service/main.tf [hunk 9]
```diff
@@ -1,15 +1,23 @@
...
+
+locals {
+  random_suffix = random_id.random_suffix.b64_url
+}
+
 module "mock_team" {
   source = "../../modules/pagerduty-team"
 
-  name        = "team - ${var.name}"
+  name        = "Team-${var.name}-${local.random_suffix}"
```
Comment (jai) [inline examples/pagerduty-business-service/main.tf:12]: capitalized on purpose?

Comment (bhargavms) [reply examples/pagerduty-business-service/main.tf:12]: Yes

File: examples/pagerduty-service/main.tf (hunk 7 above)
Comment (bhargavms) [reply examples/pagerduty-service/main.tf:14]: Yes as discussed

File: examples/pagerduty-business-service/main.tf (hunk 9 above)
Comment (bhargavms) [reply examples/pagerduty-business-service/main.tf:12]: Felt that team name deserved to be capitalized lol

File: modules/pagerduty-service-integration-email/generic-email-inbound-integration.tf [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  service           = var.service_id
+  integration_email = var.integration_email
+  # hardcoded now since we don't support multiple email filters.
+  email_filter_mode = "and-rules-email"
+  # hardcoded because this is a generic_email_inbound_integration module.
+  type                    = "generic_email_inbound_integration"
+  email_incident_creation = var.email_incident_creation
+
+  email_filter {
+    # v1 we don't expose this yet.
```
Comment (jai) [inline modules/pagerduty-service-integration-email/generic-email-inbound-integration.tf:12]: comment is not so clear. Better to say:

"In the initial implementation of this module, we hardcode this value..."
"Future enhancement: expose this attribute as a variable..."

File: modules/pagerduty-service-integration-email/generic-email-inbound-integration.tf [hunk 11]
```diff
@@ -0,0 +1,21 @@
+resource "pagerduty_service_integration" "generic_email_inbound_integration" {
+  name              = var.name
+  service           = var.service_id
+  integration_email = var.integration_email
+  # hardcoded now since we don't support multiple email filters.
```
Comment (jai) [inline modules/pagerduty-service-integration-email/generic-email-inbound-integration.tf:5]: not clear what this means

File: modules/pagerduty-service-integration-email/README.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
+## PagerDuty service integration - Email
```
Comment (jai) [inline modules/pagerduty-service-integration-email/README.md:1]:
```suggestion
## PagerDuty Email Service Integration
```

File: modules/pagerduty-service-integration-email/README.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+A `pagerduty-service-integration-email` is a [`pagerduty_service_integration`](https://registry.terraform.io/providers/PagerDuty/pagerduty/latest/docs/resources/service_integration) of the type `generic_email_inbound_integration` that belongs to a service.
```
Comment (jai) [inline modules/pagerduty-service-integration-email/README.md:3]:
```suggestion
A `pagerduty-service-integration-email` is a [`pagerduty_service_integration`](https://registry.terraform.io/providers/PagerDuty/pagerduty/latest/docs/resources/service_integration) of the type `generic_email_inbound_integration`. Use this module to trigger incidents via email messages.
```

**Review by jai — APPROVED** [review]
left some comments, some I think you'll want to fix, others can be left.

Ideally for things to be fixed later we create a ticket and link to this repo somehow - I think we have a linear label for this repo, maybe not...you can figure it out though.

Great work!

File: modules/pagerduty-service-integration-email/generic-email-inbound-integration.tf (hunk 11 above)
Comment (bhargavms) [reply modules/pagerduty-service-integration-email/generic-email-inbound-integration.tf:5]: the underlying resource has support for taking in an array of email filter objects and having AND/OR between them. IMO This comment will make sense to those that look at the doc for this resource and wonder why we didn't build support for multiple email filters.

File: modules/pagerduty-service-integration-email/inputs.tf [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  description = "The ID of the service the integration should belong to."
+  type        = string
+}
+
+variable "integration_email" {
+  description = "This is the unique fully-qualified email address used for routing emails to this integration for processing."
+  type        = string
+}
+
+variable "email_incident_creation" {
```
[the remark this answers is not in the document]
Comment (bhargavms) [reply modules/pagerduty-service-integration-email/inputs.tf:16]: keeping the naming parity with the name of the actual resource, easier to onboard new devs if we stick to same naming and less time required to explain them our jargon :D

**Pull request merged**